trinity/sync/full/chain.py

```diff
@@ -85,314 +143,194 @@ class FastChainSyncer(BaseHeaderChainSyncer):
+                    len(all_headers),
+                    all_headers[0],
+                    all_headers[-1],
                 )
-            batch_size = math.ceil(len(headers) / len(peers))
-            batches = tuple(partition_all(batch_size, headers))
-
-            # issue requests to all of the peers and wait for all of them to respond.
-            requests = tuple(
-                self._get_block_bodies(peer, batch)
-                for peer, batch
-                in zip(peers, batches)
-            )
-            responses = await self.wait(asyncio.gather(
-                *requests,
-                loop=self.get_event_loop(),
-            ))
-
-            # extract the returned block body data and the headers for which we
-            # are still missing block bodies.
-            all_block_body_bundles, all_missing_headers = zip(*responses)
-
-            for (body, (tx_root, trie_data_dict), uncles_hash) in concat(all_block_body_bundles):
-                await self.wait(self.db.coro_persist_trie_data_dict(trie_data_dict))
+                completed_headers = non_trivial_headers
+            else:
+                bundles, completed_headers = await self._get_block_bodies(peer, non_trivial_headers)
+                await self._block_body_bundle_processing(bundles)
 
-            block_bodies_by_key = merge(block_bodies_by_key, {
-                (transaction_root, uncles_hash): block_body
-                for block_body, (transaction_root, trie_dict_data), uncles_hash
-                in concat(all_block_body_bundles)
-            })
-            headers = tuple(concat(all_missing_headers))
+            self._mark_body_download_complete(batch_id, completed_headers + trivial_headers)
+        except BaseP2PError:
+            self._block_body_tasks.complete(batch_id, trivial_headers)
+            self.logger.debug("Problem downloading body from peer, dropping...", exc_info=True)
+        except Exception:
+            self._block_body_tasks.complete(batch_id, trivial_headers)
+            raise
+        else:
+            if len(non_trivial_headers) == 0:
+                # peer had nothing to do, so have it get back in line for processing
+                self._body_peers.put_nowait(peer)
+            elif len(completed_headers) > 0:
+                # peer completed with at least 1 result, so have it get back in line for processing
+                self._body_peers.put_nowait(peer)
+            else:
+                # peer returned no results, wait a while before trying again
+                delay = self.EMPTY_PEER_RESPONSE_PENALTY
+                self.logger.debug("Pausing %s for %.1fs, for sending 0 block bodies", peer, delay)
+                self.delayed_run_func(lambda: self._body_peers.put_nowait(peer), delay)
 
-        self.logger.debug("Got block bodies batch for %d headers", len(all_headers))
-        return block_bodies_by_key
+    def _mark_body_download_complete(
```

Comment:
Is this passthrough function just for readability?

Reply:
It's an unfortunate side-effect of the two classes being built with inheritance. They have to register that the download is complete in a different `OrderedTaskPreparation` object, with a different prerequisite.